Refactor architecture document into PAT-like protocols and entities (#86)
ietf-wg-privacypass/base-drafts: draft-ietf-privacypass-architecture.md

# Introduction

Privacy Pass is a protocol for authorization based on anonymous-credential
authentication mechanisms. Typical approaches for authorizing clients,
such as through the use of long-term cookies, are not privacy-friendly
since they allow servers to track clients across sessions and interactions.
Privacy Pass takes a different approach: instead of presenting linkable
state carrying information to servers, e.g., whether or not the client
is an authorized user or has completed some prior challenge, clients
present unlinkable proofs that attest to this information.

The most basic Privacy Pass protocol provides a set of cross-origin
authorization tokens that protect the client's anonymity during interactions
with a server. This allows clients to communicate an attestation of a
previously authenticated server action, without having to reauthenticate
manually. The tokens retain anonymity in the sense that the act of
revealing them cannot be linked back to the session where they were
initially issued.

> At a high level, Privacy Pass is composed of two protocols: issuance
> and redemption. Issuance is a protocol between a Client and two functions:
> Attester and Issuer, where the Attester and Issuer can be functions operated
> by the same protocol participant. The Issuer is responsible for issuing tokens
> in response to requests from Clients. The Attester is responsible for attesting
> properties about the Client for which tokens are issued. For example, in the
> original Privacy Pass protocol {{PPSRV}}, tokens were only issued to Clients
> that solved CAPTCHAs. In this context, the Attester attested that some client
> solved a CAPTCHA and the resulting token produced by the Issuer was proof of
> this fact. Depending on the information being attested, Attesters may also
> store state about individual Clients, such as the number of overall tokens
> issued thus far.

Reviewer: ```suggestion
At a high level, Privacy Pass is composed of two protocols: issuance
and redemption.

Issuance is a protocol between a Client and two functions: Attester and Issuer.
The Attester and Issuer can be functions operated by the same protocol participant,
but can also be implemented separately. The Issuer is responsible for issuing tokens
in response to requests from Clients. The Attester is responsible for attesting
properties about the Client for which tokens are issued. The Issuer needs to be
trusted by the server that later redeems the token. The Attester needs to either be
operated by the Issuer, or be operated by an entity the Issuer trusts to perform valid
attestation. Clients might prefer to select different Attesters, separate from the
Issuer, to be able to use preferred authentication methods or improve privacy by
not directly communicating with an Issuer. Depending on the information being attested, Attesters may also store state about individual Clients, such as the number of
overall tokens issued thus far. As an example of an Issuance protocol, in the
original Privacy Pass protocol {{PPSRV}}, tokens were only issued to Clients
that solved CAPTCHAs. In this context, the Attester attested that some client
solved a CAPTCHA and the resulting token produced by the Issuer was proof of
this fact.
```

# Introduction

Privacy Pass is a protocol for authorization based on anonymous-credential
authentication mechanisms. Typical approaches for authorizing clients,
such as through the use of long-term cookies, are not privacy-friendly
since they allow servers to track clients across sessions and interactions.
Privacy Pass takes a different approach: instead of presenting linkable
state carrying information to servers, e.g., whether or not the client
is an authorized user or has completed some prior challenge, clients
present unlinkable proofs that attest to this information.

> The most basic Privacy Pass protocol provides a set of cross-origin

Reviewer 2: cross-origin is a little odd now, since the per-origin tokens aren't quite "cross-origin" in the traditional sense versus cross between Origin and Issuer.

Author: Hmm, I don't follow. Can you elaborate?

performed by the Issuer or by an Attester that is trusted by the Issuer.
Clients might prefer to select different Attesters, separate from the Issuer,
to be able to use preferred authentication methods or improve privacy by not
directly communicating with an Issuer. Depending on the attestation,
Attesters can store state about a Client, such as the number of overall tokens
issued thus far. As an example of an Issuance protocol, in the original Privacy
Pass protocol {{PPSRV}}, tokens were only issued to Clients that solved
CAPTCHAs. In this context, the Attester attested that some client solved a
CAPTCHA and the resulting token produced by the Issuer was proof of this fact.

The redemption protocol runs between Client and Origin (server). It allows
Origins to challenge Clients to present one or more tokens for authorization.
Depending on the type of token, e.g., whether or not it is cross-origin
or per-origin, and whether or not it can be cached, the Client either presents
a previously obtained token or invokes the issuance protocol to acquire one
for authorization.

The issuance and redemption protocols operate in concert as shown in
the figure below.

~~~

>  Client        Attester          Issuer          Origin

Reviewer 2: Maybe worth moving Origin to the left to match the other figures?

Author: Sure, I'll play with it and see what works.

the parametrization in {{parametrization}}.

The mechanism by which clients obtain the Issuer public key is not specified.
Clients may be configured with this key or they may discover it via some other
form. See {{?CONSISTENCY=I-D.wood-key-consistency}}.

Depending on the use case, issuance may require some form of Client
anonymization service similar to an IP-hiding proxy so that Issuers cannot
learn information about Clients. This can be provided by an explicit
participant in the issuance protocol, or it can be provided via external means,
e.g., through the use of an IP-hiding proxy service like Tor. In general,
Clients SHOULD minimize or remove identifying information where possible when
invoking the issuance protocol.

Issuers MUST NOT issue tokens for Clients through untrusted Attesters. This is
important because the Attester's role is to vouch for trust in
privacy-sensitive Client information, such as account identifiers or IP address
information, to the Issuer. Tokens produced by an Issuer that admits issuance
for any type of attestation cannot be relied on for any specific property.
See {{attester-role}} for more details.

> ### Attester Role

Reviewer 2: Maybe later on, but should it be explicitly called out that some forms of the protocol might collapse this into another role?

Attestation is an important part of the issuance protocol. Attestation is the
process by which an Attester bears witness to, confirms, or authenticates a
Client so as to verify a property about the Client that is required for
Issuance. Examples of attestation properties include, though are not limited
to:

- Capable of solving a CAPTCHA. Clients that solve CAPTCHA challenges can attest
  to this capability for the purposes of being ruled out as a bot or otherwise
  automated Client.

> - Client state. Clients can be associated with state and the attestor can

Reviewer 2: Maybe worth splitting out client identity and client issuance state?

  attest to this state. Examples of state include the number of issuance
  protocol invocations, the client's geographic region, and whether the
  client has a valid application-layer account.
- Trusted device. Some Clients run on trusted hardware that are capable of
  producing device-level attestation statements.

Each of these attestation types have different security properties. For
example, attesting to having a valid account is different from attesting to be
running on trusted hardware. In general, Attesters should accept a limited form
of attestation formats.

Each attestation format also has an impact on the overall system privacy. For
example, the number of users in possession of a single class of trusted device
might be lesser than the number of users that can solve CAPTCHAs. Similarly,
requiring a conjunction of attestation types could decrease the overall
anonymity set size. For example, the number of Clients that have solved a
CAPTCHA in the past day, have a valid account, and are running on a trusted

> device is lesser thant he number of Clients that have solved a CAPTCHA in the

Reviewer 2: nit: than the

# Introduction

Privacy Pass is a protocol for authorization based on anonymous-credential
authentication mechanisms. Typical approaches for authorizing clients,
such as through the use of long-term cookies, are not privacy friendly
since they allow servers to track clients across sessions and interactions.
Privacy Pass takes a different approach: instead of presenting linkable
state carrying information to servers, e.g., whether or not the client
is an authorized user or has completed some prior challenge, clients
present unlinkable proofs that attest to this information.

> The original Privacy Pass protocol provides a set of cross-origin

Reviewer: Does this need to say "original"?

```suggestion
The most basic Privacy Pass protocol provides a set of cross-origin
```

authorization tokens that protect the client's anonymity during interactions
with a server. This allows clients to communicate an attestation of a
previously authenticated server action, without having to reauthenticate
manually. The tokens retain anonymity in the sense that the act of
revealing them cannot be linked back to the session where they were
initially issued.

At a high level, Privacy Pass is composed of two protocols: issuance
and redemption. Issuance is a protocol between three logical entities:

> Client, Mediator, and Issuer. The Issuer is responsible for issuing

Reviewer: We can also bike shed on the "mediator" name. I do like the "attest" verb you are using elsewhere. "Attestor"?

For example, in the original Privacy Pass protocol {{PPSRV}}, tokens were
only issued to Clients that solved CAPTCHAs. In this context, the Mediator
attested that some client solved a CAPTCHA and the resulting token
produced by the Issuer was proof of this fact. Depending on the information
being attested, Mediators may also store state about individual Clients,
such as the number of overall tokens issued thus far.

The redemption protocol runs between Client and Origin (server). It allows
Origins to challenge Clients to present one or more tokens for authorization.
Depending on the type of token, e.g., whether or not it is cross-origin
or per-origin, and whether or not it can be cached, the Client either presents
a previously obtained token or invokes the issuance protocol to acquire one
for authorization.

The issuance and redemption protocols operate in concert as shown in
the figure below.

~~~
 Client        Mediator          Issuer          Origin

    <---------------------------------------- Challenge \
                                                        |
+--------------------------------------------\          |

> | TokenRequest --->                          |          |

Reviewer: Perhaps the Client -> Mediator flow should indicate a first authentication/validation step?

Reviewer: I like the later diagram that has "attest"

|                    TokenRequest --->       |          |
|                                 (validate) | Issuance | Redemption
|                                 (evaluate) |   Flow   |   Flow
|                    <--- TokenResponse      |          |
|   <--- TokenResponse                       |          |
---------------------------------------------/          |

>                                                         |

Reviewer 3: I definitely like this new figure better.

     Response -------------------------------------- >  /
~~~
{: #fig-overview title=" Privacy Pass Architectural Components"}

This document describes requirements for both issuance and redemption
protocols. This document also describes ecosystem considerations that
impact the stated privacy and security guarantees of the protocol.
For instance, the number of servers issuing Privacy Pass tokens, along
with the number of registered clients, determines the anonymity set of
each individual client. Moreover, this can be influenced by other factors,
such as the key rotation policy used by each server. There are also client
behavior patterns that can reduce the effective security of the server.

> Additional considerations include:

Reviewer 3: This and the following points all seem unnecessary in the Intro, given the preceding sentence that says that the document will describe ecosystem considerations.

initially issued.

At a high level, Privacy Pass is composed of two protocols: issuance
and redemption. Issuance is a protocol between a Client and two functions:
Attester and Issuer. The Attester and Issuer can be functions operated by the
same protocol participant, but can also be implemented separately. The Issuer
is responsible for issuing tokens in response to requests from Clients. The
Attester is responsible for attesting properties about the Client for which

> tokens are issued. The Issuer needs to be trusted by the server that later

Author: The descriptions of these roles _are_ right after they're introduced, so I'm not sure what this looking for here. Leave a suggestion?

redeems the token. The Attester needs to either be operated by the Issuer,
or be operated by an entity the Issuer trusts to perform valid attestation.
Clients might prefer to select different Attesters, separate from the Issuer,
to be able to use preferred authentication methods or improve privacy by not

> directly communicating with an Issuer. Depending on the information being

Author: As discussed, I don't think we should describe the Attester as a proxy.

> directly communicating with an Issuer. Depending on the information being
> attested, Attesters may also store state about individual Clients, such as the

Reviewer 3: ```suggestion
directly communicating with an Issuer. Depending on the attestation,
Attesters can store state about a Client, such as the
```

- Client: An entity that seeks authorization to an Origin.
- Origin: An entity that challenges Clients for tokens.
- Issuer: An entity that issues tokens to Clients for properties

>   attested by the Attester.

Reviewer 3: ```suggestion
  attested to by the Attester.
```

- Attester: An entity that attests to properties of Client for the

>   purposes of token issuance.

Reviewer 3: ```suggestion
- Attester: An entity that attests to properties of Client for the
  purposes of token issuance.
```
```suggestion
- Attester: An entity that attests to the properties of a Client required by an Issuer
  for token issuance.
```

The redemption protocol is a simple challenge-response based authorization
protocol between Client and Origin. Origins prompt Clients with a token
challenge and, if possible, Clients present a valid token for the challenge
in response. The challenge controls the type of token that the Origin will
accept for the given resource. As described in [http-auth-doc], there are
a number of ways in which the token may vary, including:

- Issuance protocol. The token identifies the type of issuance protocol
  required for producing the token. Different issuance protocols have different
  security properties, e.g., some issuance protocols may produce tokens that
  are publicly verifiable, whereas others may not have this property.
- Interactive or non-interactive. Tokens can either be interactive or not.
  An interactive token is one which requires a freshly issued token based
  on the challenge, whereas a non-interactive token can be issued proactively
  and cached for future use.
- Per-domain or cross-domain. Tokens can be constrained to the Origin for
  which the challenge originated, or can be used across Origins.
- Trusted issuer. Tokens identify which issuers are trusted for a given

>   issuance protocol.

Author: Yep, that's better 👍

parties that produces a valid authenticator over the Client's input, subject
to the following security requirements.

1. Unconditional input secrecy. The issuance protocol MUST NOT reveal anything
about the Client's private input, including the challenge and nonce. The
issuance protocol can reveal the Issuer public key for the purposes of
determining which private key to use in producing the issuance protocol. A
result of this property is that the redemption flow is unlinkable from the
issuance flow.
1. One-more forgery security. The issuance protocol MUST NOT allow malicious
Clients to forge tokens without interacting with the Issuer directly.
1. Concurrent security. The issuance protocol MUST be safe to run concurrently
with arbitrarily many Clients.

Each Issuance protocol MUST come with a detailed analysis of the privacy impacts
of the protocol, why these impacts are justified, and guidelines on changes to
the parametrization in {{parametrization}}.

The mechanism by which clients obtain the Issuer public key is not specified.
Clients may be configured with this key or they may discover it via some other
form. See {{?CONSISTENCY=I-D.wood-key-consistency}}.

> Depending on the use case, issuance may requrie some form of Client

Reviewer 3: ```suggestion
Depending on the use case, issuance may require some form of Client
```

anonymization service similar to an IP-hiding proxy so that Issuers cannot
learn information about Clients. This can be provided by an explicit
participant in the issuance protocol, or it can be provided via external means,
e.g., through the use of an IP-hiding proxy service like Tor. In general,
Clients SHOULD minimize or remove identifying information where possible when
invoking the issuance protocol.

Issuers MUST NOT issue tokens for Clients through untrusted Mediators. This is
important because the Mediator's role is to vouch for trust in
privacy-sensitive Client information, such as account identifiers or IP address
information, to the Issuer. Tokens produced by an Issuer that admits issuance
for any type of mediation cannot be relied on for any specific property.
See {{attestation}} for more details.

### Mediator Role

Mediation is an important part of the issuance protocol. Mediaton involves
two high level functions: attestation and accounting. These are described in
more detail below.

#### Attestation

> Attestation is the process by which the Client's bear witness, confirm, or

Reviewer 3: ```suggestion
Attestation is the process by which the Clients bear witness, confirm, or
```

want to limit the number of tokens issued to a single Client over the course
of a day. If the Issuer does not learn information about the Client, then the
Issuer cannot enforce this limit on a per-Client basis. The Issuer could,
however, enforce global rate limits, but these can be abused by malicious
Clients at the determent of honest Clients. Thus, the task of accounting
falls on the Mediator, since it is entrusted with some amount of
privacy-sensitive Client information.

### Issuer Role

Issuers MUST be uniquely identifiable by all Clients with a consistent
identifier. In a web context, this identifier might be the Issuer host name.
As discussed later in {{privacy}}, ecosystems that admit a large number of
Issuers can lead to privacy concerns for the Clients in the ecosystem.
Therefore, in practice, the number of Issuers should be bounded. The actual
Issuers can be replaced with different Issuers as long as the total never
exceeds these bounds. Moreover, Issuer replacements also have an effect on
client anonymity that is similar to when a key rotation occurs. See {{privacy}}
for more details about maintaining privacy with multiple servers.

#### Key Management

> To facilitate issuance, the Issuer MUST hold a Privacy Pass key pair at any

Reviewer 3: This is the first time "Privacy Pass key pair" is mentioned. Given that this key pair is described below, perhaps rename the section "Privacy Pass Keys"?

Also, server public key should be Issuer public key, right?

given time. The server public key MUST be made available to all Clients in
such a way that key rotations and other updates are publicly visible. The

> server MAY also require additional state for ensuring this.  The key material

Reviewer 3: unclear: for ensuring what?

Author: I'm not sure -- this is leftover text, so I'll just remove it.

Clients may be configured with this key or they may discover it via some other
form. See {{?CONSISTENCY=I-D.wood-key-consistency}}.

Depending on the use case, issuance may require some form of Client
anonymization service similar to an IP-hiding proxy so that Issuers cannot
learn information about Clients. This can be provided by an explicit
participant in the issuance protocol, or it can be provided via external means,
e.g., through the use of an IP-hiding proxy service like Tor. In general,
Clients SHOULD minimize or remove identifying information where possible when
invoking the issuance protocol.

Issuers MUST NOT issue tokens for Clients through untrusted Attesters. This is
important because the Attester's role is to vouch for trust in
privacy-sensitive Client information, such as account identifiers or IP address
information, to the Issuer. Tokens produced by an Issuer that admits issuance
for any type of attestation cannot be relied on for any specific property.
See {{attester-role}} for more details.

### Attester Role

Attestation is an important part of the issuance protocol. Attestation is the

> process by which the Clients bear witness, confirm, or authenticate so as to

Reviewer 3: `bear witness` seems strange. What can a Client bearing witness to about itself? Maybe this should be `by which an Attester bears witness to, confirms, or authenticates a Client so as to verify a property about the Client that is required for Issuance.`

Author: Yeah, that's better!

vector.

Private metadata is that which clients cannot observe as part of the token
issuance flow. Such instantiations may be built on the Private Metadata Bit
construction from Kreuter et al. {{?KLOR20=DOI.10.1007/978-3-030-56784-2_11}}
or the attribute-based VOPRF from Huang et al. {{HIJK21}}.

Metadata may also be arbitrarily long or bounded in length. The amount of
permitted metadata may be determined by application or by the underlying
cryptographic protocol.

### Extensibility {#extensions}

The Privacy Pass protocol and ecosystem are both intended to be

> receptive to extensions that expand the current set of functionality.

Reviewer 3: ```suggestion
receptive to extensions that expand the current set of functionalities.
```

All extensions to the Privacy Pass protocol SHOULD be specified as separate

> documents that modify the content of this document in some way.

Reviewer 3: I don't understand this requirement -- why SHOULD extensions modify this document? Also, there's a Redemption protocol and an Issuance protocol -- there is no mention of the Privacy Pass protocol so far in this doc.

Author: Rephrased!

Any such extension SHOULD come with a detailed analysis of the privacy
impacts of the extension, why these impacts are justified, and guidelines
on changes to the parametrization in {{parametrization}}.
Any extension to the Privacy Pass protocol MUST adhere to the guidelines
specified in {{issuer-role}} for managing Issuer public key data.

# Privacy considerations {#privacy}

The goal of the Privacy Pass ecosystem is to construct an environment

> that can easily measure (and maximize) the relative anonymity of any

Reviewer 3: ```suggestion
that can easily measure and maximize the relative anonymity of any
```

similarly to introducing more servers (though with more potential
targeting).

For this reason, the amount of metadata used by an server in creating
redemption tokens must be taken into account -- together with the bits
of information that server's may learn about Clients otherwise. Since this
metadata may be useful for practical deployments of Privacy Pass, servers
must balance this against the reduction in Client privacy. In general,
servers should permit no more than 32 bits of metadata, as this can
uniquely identify each possible user. We discuss this more in
{{parametrization}}.

## Issuer key rotation

Techniques to introduce Client "segregation" can be used to reduce
Client anonymity. Such techniques are closely linked to the type of key
schedule that is used by the Issuer. When an Issuer rotates their key,
any Client that invokes the issuance protocol in this key cycle will be
part of a group of possible clients owning valid tokens for this key. To
mechanize this attack strategy, an Issuer could introduce a key rotation
policy that forces Clients into small key cycles. Thus, reducing the
size of the anonymity set for these Clients.

> It is RECOMMENDED that Issuers should only invoke key rotation for fairly
> large periods of time such as between 1 and 12 weeks. Key rotations

Reviewer 3: ```suggestion
Issuers SHOULD invoke key rotation for
a period of time between 1 and 12 weeks. Key rotations
```
RECOMMENDED and SHOULD have the same semantic, and SHOULD seems more direct here. Also, if there's a SHOULD, the recommendation should be clear.

represent a trade-off between Client privacy and continued Issuer
security. Therefore, it is still important that key rotations occur on a

> fairly regular cycle to reduce the harmfulness of a Issuer key

Reviewer 3: ```suggestion
regular cycle to reduce the harmfulness of an Issuer key
```

compromise.

> With a large number of Clients, a week gives a fairly large window for
> Clients to participate in the Privacy Pass protocol and thus enjoy the
> anonymity guarantees of being part of a larger group. A low ceiling of

Reviewer 3: ```suggestion
With a large number of Clients, a minimum of one week gives a large enough window for
Clients to participate in the Issuance protocol and thus enjoy the
anonymity guarantees of being part of a larger group. A maximum of
```

> 12 weeks prevents a key compromise from being too destructive. If an Issuer

Reviewer 3: ```suggestion
12 weeks limits the damage caused by a key compromise. If an Issuer
```

realizes that a key compromise has occurred then the Issuer should
generate a new key and make it available to Clients. If possible, it should
invoke any revocation procedures that may apply for the old key.

## Large numbers of Issuers {#servers}

Similarly to the Issuer rotation dynamic that is raised above, if there
are a large number of Issuers then segregation can occur. For example,
if Clients obtain tokens from many Issuers, and Origins later challenge

> Client for a token from each Issuer, Origins can learn information about

Reviewer 3: ```suggestion
a Client for a token from each Issuer, Origins can learn information about
```

an Issuer, even if the tokens fail later verification checks. This is
because tokens effectively represent a form of currency that they can
later redeem for some sort of benefit. The verification checks that are
put in place are there to ensure that the Client does not sacrifice
their anonymity. However, a Client may judge the "monetary" benefit of
owning tokens to be greater than their own privacy.

Firstly, a Client behaving in this way would not be compliant with the
protocol, as laid out in {{?I-D.ietf-privacypass-protocol}}.

Secondly, acting in this way only affects the privacy of the immediate
Client. There is an exception if a large number of Clients colluded to
accept bad data, then any Client that didn't accept would be part of a
smaller anonymity set. However, such a situation would be identical to
the situation where the total number of Clients in the ecosystem is
small. Therefore, the reduction in the size of the anonymity set would
be equivalent; see {{servers}} for more details.

# Security considerations {#security}

We present a number of security considerations that prevent malicious
Clients from abusing the protocol.

## Double-spend protection {#double-spend}

> When applicable, all Origins SHOULD implement a robust storage-query

Reviewer: I think we should be clear that this is only need for non-interactive passes.

mechanism for checking that tokens sent by clients have not been spent before.
Such tokens only need to be checked for each Origin individually. But all
Origins must perform global double-spend checks to avoid clients from
exploiting the possibility of spending tokens more than once against
distributed token checking systems. For the same reason, the global data
storage must have quick update times. While an update is occurring it
may be possible for a malicious client to spend a token more than once.

## Token exhaustion

> When a Client holds tokens for an Issuer, it is possible for any

Reviewer: We should mention tokens being scoped to an origin here, or being interactive, as a way to avoid this.

verifier to invoke that client to redeem tokens for that Issuer. This
can lead to an attack where a malicious verifier can force a Client to
spend all of their tokens for a given Issuer. To prevent this from
happening, methods should be put into place to prevent many tokens from
being redeemed at once.

For example, it may be possible to cache a redemption for the entity
that is invoking a token redemption. If the verifier requests more
tokens then the Client simply returns the cached token that it returned
previously. This could also be handled by simply not redeeming any
tokens for verification if a redemption had already occurred in a given
time window.

In AV, the Client instead caches the SRR that it received in the
asynchronous redemption exchange with the Issuer. If the same verifier
attempts another redemption request, then the client simply returns the
cached SRR. The SRRs can be revoked by the Issuer, if need be, by
providing an expiry date or by signaling that records from a particular

> window need to be refreshed.

Reviewer: AV and SRR are no longer defined in the document.

value of I should be 3.

If the value of U is much bigger (e.g. 5 million) then this would permit
I = (log_2(5000000/5000)-1)/2 ~= 4 Issuers.

## Allowing more Issuers

> Using the recommendations in {{more-servers}}, it is possible to

Reviewer: I do like the approach of having clients limit the number of non-interactive cross-origin tokens they're willing to maintain from different Issuers.
